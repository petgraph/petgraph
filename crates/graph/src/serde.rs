//! Serde support for `Graph` and `StableGraph`
//!
//! ## Serialization Format
//!
//! The serialization format is as follows, in pseudocode:
//!
//! ```text
//! Graph {
//!     nodes: [N],
//!     node_holes: [NodeIndex<Ix>],
//!     edge_property: EdgeProperty,
//!     edges: [Option<(NodeIndex<Ix>, NodeIndex<Ix>, E)>]
//! }
//! ```
//!
//! Node indices are serialized as integers and are fixed size for binary formats, so the Ix
//! parameter matters there.
//!
//! A stable graph serialization that obeys these restrictions (effectively, it has no interior
//! vacancies) can de deserialized as a [`Graph`].
//!
//! ## [`Graph`] Restrictions
//!
//! `node_holes` is always empty and `edges` are always `Some`.

use alloc::{string::ToString, vec::Vec};
use core::{
    fmt::{Display, Formatter},
    marker::PhantomData,
};

use funty::Integral;
use petgraph_core::{edge::EdgeType, index::IndexType};
use serde::{
    de::Error as _,
    ser::{SerializeStruct, SerializeTuple},
    Deserialize, Deserializer, Serialize, Serializer,
};

use super::{EdgeIndex, NodeIndex};
use crate::{Edge, Graph, Node};

#[derive(Debug)]
pub enum InvalidError {
    Node {
        index: usize,
        length: usize,
    },
    Hole {
        index: usize,
    },
    Length {
        type_: &'static str,
        length: usize,
        max: usize,
    },
    Direction {
        expected: EdgeProperty,
        received: EdgeProperty,
    },
}

impl Display for InvalidError {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Node { index, length } => f.write_fmt(format_args!(
                "invalid value: node index `{index}` does not exist in graph with length \
                 `{length}`",
            )),
            Self::Hole { index } => f.write_fmt(format_args!(
                "invalid value: node hole `{index}` is not allowed.",
            )),
            Self::Length { type_, length, max } => f.write_fmt(format_args!(
                "invalid value: {type_} length `{length}` exceeds maximum of `{max}`",
            )),
            Self::Direction { expected, received } => f.write_fmt(format_args!(
                "invalid value: expected {expected} graph, but received {received} graph",
            )),
        }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EdgeProperty {
    Undirected,
    Directed,
}

impl Display for EdgeProperty {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Undirected => f.write_str("undirected"),
            Self::Directed => f.write_str("directed"),
        }
    }
}

impl EdgeProperty {
    pub(crate) fn from_type<T>() -> Self
    where
        T: EdgeType,
    {
        if T::is_directed() {
            Self::Directed
        } else {
            Self::Undirected
        }
    }
}

impl<Ix> Serialize for NodeIndex<Ix>
where
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<Ix> Serialize for EdgeIndex<Ix>
where
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<E, Ix> Serialize for Edge<E, Ix>
where
    E: Serialize,
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut tuple = serializer.serialize_tuple(3)?;

        tuple.serialize_element(&self.source())?;
        tuple.serialize_element(&self.target())?;
        tuple.serialize_element(&self.weight)?;

        tuple.end()
    }
}

impl<N, Ix> Serialize for Node<N, Ix>
where
    N: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.weight.serialize(serializer)
    }
}

impl<N, E, Ty, Ix> Serialize for Graph<N, E, Ty, Ix>
where
    N: Serialize,
    E: Serialize,
    Ty: EdgeType,
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut struct_ = serializer.serialize_struct("Graph", 4)?;

        struct_.serialize_field("nodes", &self.nodes)?;
        struct_.serialize_field("node_holes", &[(); 0])?;
        struct_.serialize_field("edge_property", &EdgeProperty::from_type::<Ty>())?;
        struct_.serialize_field("edges", &self.edges)?;

        struct_.end()
    }
}

impl<'de, Ix> Deserialize<'de> for NodeIndex<Ix>
where
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ix::deserialize(deserializer).map(NodeIndex)
    }
}

impl<'de, Ix> Deserialize<'de> for EdgeIndex<Ix>
where
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ix::deserialize(deserializer).map(EdgeIndex)
    }
}

#[derive(Deserialize)]
#[serde(rename = "Edge")]
struct EdgeRemote<E, Ix: IndexType>(NodeIndex<Ix>, NodeIndex<Ix>, E);

impl<'de, E, Ix> Deserialize<'de> for Edge<E, Ix>
where
    E: Deserialize<'de>,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let EdgeRemote(source, target, weight) = EdgeRemote::deserialize(deserializer)?;

        Ok(Edge {
            weight,
            node: [source, target],
            next: [EdgeIndex::end(); 2],
        })
    }
}

#[derive(Deserialize)]
#[serde(rename = "Node")]
struct NodeRemote<N>(N);

impl<'de, N, Ix> Deserialize<'de> for Node<N, Ix>
where
    N: Deserialize<'de>,
    Ix: IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let NodeRemote(weight) = NodeRemote::deserialize(deserializer)?;

        Ok(Node {
            weight,
            next: [EdgeIndex::end(); 2],
        })
    }
}

#[derive(Deserialize)]
#[serde(rename = "Graph")]
struct RemoteGraph<N, E, Ix: IndexType> {
    nodes: Vec<Node<N, Ix>>,
    // always empty per serialization format
    #[serde(default)]
    node_holes: [(); 0],
    edge_property: EdgeProperty,
    edges: Vec<Edge<E, Ix>>,
}

impl<'de, N, E, Ty, Ix> Deserialize<'de> for Graph<N, E, Ty, Ix>
where
    N: Deserialize<'de>,
    E: Deserialize<'de>,
    Ty: EdgeType,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let RemoteGraph {
            nodes,
            edge_property,
            edges,
            ..
        } = RemoteGraph::deserialize(deserializer)?;

        if nodes.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "node",
                length: nodes.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        if edges.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "edge",
                length: edges.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        let expected = EdgeProperty::from_type::<Ty>();
        if edge_property != expected {
            return Err(InvalidError::Direction {
                expected,
                received: edge_property,
            })
            .map_err(D::Error::custom);
        }

        let mut this = Self {
            nodes,
            edges,
            ty: PhantomData,
        };

        let node_count = this.node_count();
        this.link_edges()
            .map_err(|i| InvalidError::Node {
                index: i.index(),
                length: node_count,
            })
            .map_err(D::Error::custom)?;

        Ok(this)
    }
}
