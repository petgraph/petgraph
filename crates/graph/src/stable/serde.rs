use alloc::vec::Vec;
use core::{cell::Cell, marker::PhantomData};

use funty::Integral;
use petgraph_core::deprecated::{
    index::IndexType,
    visit::{EdgeIndexable, NodeIndexable},
};
use serde::{
    de::Error as _,
    ser::{SerializeSeq, SerializeStruct},
    Deserialize, Deserializer, Serialize, Serializer,
};

use crate::{
    serde::{EdgeProperty, InvalidError},
    stable::StableGraph,
    Edge, EdgeIndex, EdgeType, Graph, Node, NodeIndex,
};

struct SerializeIter<T> {
    len: usize,
    iter: Cell<Option<T>>,
}

impl<T> SerializeIter<T> {
    const fn new(len: usize, iter: T) -> Self {
        Self {
            len,
            iter: Cell::new(Some(iter)),
        }
    }
}

impl<T> Serialize for SerializeIter<T>
where
    T: Iterator,
    T::Item: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let iter = self.iter.take().expect("SerializeIter already serialized");

        let mut seq = serializer.serialize_seq(Some(self.len))?;
        let mut count = 0;
        for item in iter {
            seq.serialize_element(&item)?;
            count += 1;
        }

        debug_assert_eq!(count, self.len, "SerializeIter length mismatch");
        seq.end()
    }
}

impl<N, E, Ty, Ix> Serialize for StableGraph<N, E, Ty, Ix>
where
    N: Serialize,
    E: Serialize,
    Ty: EdgeType,
    Ix: Serialize + IndexType,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let nodes = &self.raw_nodes()[..self.node_bound()];
        let node_count = self.node_count();
        let hole_count = nodes.len() - node_count;
        let edges = &self.raw_edges()[..self.edge_bound()];

        let mut struct_ = serializer.serialize_struct("StableGraph", 4)?;

        struct_.serialize_field(
            "nodes",
            &SerializeIter::new(
                node_count,
                nodes.iter().filter_map(|node| node.weight.as_ref()),
            ),
        )?;
        struct_.serialize_field(
            "node_holes",
            &SerializeIter::new(
                hole_count,
                nodes.iter().enumerate().filter_map(|(index, node)| {
                    node.weight.is_none().then(|| NodeIndex::<Ix>::new(index))
                }),
            ),
        )?;
        struct_.serialize_field("edge_property", &EdgeProperty::from_type::<Ty>())?;

        // Convert from `Edge<Option<E, Ix>>` to `Option<Edge<E, Ix>>`
        let edges_len = edges.len();
        struct_.serialize_field(
            "edges",
            &SerializeIter::new(
                edges_len,
                edges.iter().map(|edge| {
                    edge.weight.as_ref().map(|weight| Edge {
                        weight,
                        next: edge.next,
                        node: edge.node,
                    })
                }),
            ),
        )?;

        struct_.end()
    }
}

#[derive(Deserialize)]
#[serde(rename = "Graph")]
struct RemoteStableGraph<N, E, Ix: IndexType> {
    nodes: Vec<Node<N, Ix>>,
    #[serde(default)]
    node_holes: Vec<NodeIndex<Ix>>,
    edge_property: EdgeProperty,
    edges: Vec<Option<Edge<E, Ix>>>,
}

impl<'de, N, E, Ty, Ix> Deserialize<'de> for StableGraph<N, E, Ty, Ix>
where
    N: Deserialize<'de>,
    E: Deserialize<'de>,
    Ty: EdgeType,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let RemoteStableGraph {
            nodes,
            node_holes,
            edge_property,
            edges,
        } = RemoteStableGraph::<N, E, Ix>::deserialize(deserializer)?;

        // convert from `Option<Edge<E, Ix>>` to `Edge<Option<E>, Ix>`
        let edges = edges
            .into_iter()
            .map(|edge| match edge {
                Some(Edge { weight, next, node }) => Edge {
                    weight: Some(weight),
                    next,
                    node,
                },
                None => Edge {
                    weight: None,
                    next: [EdgeIndex::end(); 2],
                    node: [NodeIndex::end(); 2],
                },
            })
            .collect::<Vec<_>>();

        if edges.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "edge",
                length: edges.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        let expected = EdgeProperty::from_type::<Ty>();
        if edge_property != expected {
            return Err(InvalidError::Direction {
                expected,
                received: edge_property,
            })
            .map_err(D::Error::custom);
        }

        let total_nodes = nodes.len() + node_holes.len();
        let mut compact_nodes = nodes.into_iter().map(|node| Node {
            weight: Some(node.weight),
            next: node.next,
        });

        let mut nodes = Vec::with_capacity(total_nodes);

        let mut node_pos = 0;
        for position in &node_holes {
            let position = position.index();

            if !(node_pos..total_nodes).contains(&position) {
                return Err(InvalidError::Hole { index: position }).map_err(D::Error::custom);
            }

            nodes.extend(compact_nodes.by_ref().take(position - node_pos));
            nodes.push(Node {
                weight: None,
                next: [EdgeIndex::end(); 2],
            });

            node_pos = position + 1;
            debug_assert_eq!(nodes.len(), node_pos);
        }

        nodes.extend(compact_nodes);

        if nodes.len() >= <Ix as Integral>::MAX.as_usize() {
            return Err(InvalidError::Length {
                type_: "node",
                length: nodes.len(),
                max: Ix::MAX.as_usize(),
            })
            .map_err(D::Error::custom);
        }

        let node_bound = nodes.len();
        let mut stable_graph = Self {
            g: Graph {
                nodes,
                edges,
                ty: PhantomData,
            },
            node_count: 0,
            edge_count: 0,
            free_edge: EdgeIndex::end(),
            free_node: NodeIndex::end(),
        };

        stable_graph
            .link_edges()
            .map_err(|i| InvalidError::Node {
                index: i.index(),
                length: node_bound,
            })
            .map_err(D::Error::custom)?;

        Ok(stable_graph)
    }
}

#[cfg(test)]
mod tests {
    use alloc::vec::Vec;

    use serde::Deserialize;

    use crate::{
        node_index,
        serde::EdgeProperty,
        stable::{serde::RemoteStableGraph, StableUnGraph},
        EdgeIndex, Node,
    };

    #[test]
    fn deserialization_with_holes() {
        let mut graph = StableUnGraph::<i32, ()>::with_capacity(0, 0);

        let a = graph.add_node(0);
        let b = graph.add_node(1);
        let c = graph.add_node(2);
        let d = graph.add_node(3);
        let e = graph.add_node(4);
        let f = graph.add_node(5);
        let g = graph.add_node(6);

        graph.remove_node(a);
        graph.remove_node(c);
        graph.remove_node(d);
        // as optimization technique, we use `node_bound()` for raw_nodes() and raw_edges(), this
        // means that if we have a trailing hole it will be skipped.
        graph.remove_node(g);

        let value = serde_value::to_value(&graph).unwrap();
        let graph = StableUnGraph::<i32, ()>::deserialize(value).unwrap();

        assert_eq!(graph.node_count(), 3);
        assert_eq!(
            graph
                .raw_nodes()
                .iter()
                .map(|n| n.weight)
                .collect::<Vec<_>>(),
            vec![None, Some(1), None, None, Some(4), Some(5)],
        );
    }
}
