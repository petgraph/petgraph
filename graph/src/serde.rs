//! Serde support for `Graph` and `StableGraph`
//!
//! ## Serialization Format
//!
//! The serialization format is as follows, in pseudocode:
//!
//! ```text
//! Graph {
//!     nodes: [N],
//!     node_holes: [NodeIndex<Ix>],
//!     edge_property: EdgeProperty,
//!     edges: [Option<(NodeIndex<Ix>, NodeIndex<Ix>, E)>]
//! }
//! ```
//!
//! Node indices are serialized as integers and are fixed size for binary formats, so the Ix
//! parameter matters there.
//!
//! A stable graph serialization that obeys these restrictions (effectively, it has no interior
//! vacancies) can de deserialized as a [`Graph`].
//!
//! ## [`Graph`] Restrictions
//!
//! `node_holes` is always empty and `edges` are always `Some`.

use alloc::{string::ToString, vec::Vec};
use core::{
    fmt::{Display, Formatter},
    marker::PhantomData,
};

use funty::Integral;
use petgraph_core::{edge::EdgeType, index::IndexType};
use serde::{
    de::Error as _,
    ser::{SerializeStruct, SerializeTuple},
    Deserialize, Deserializer, Serialize, Serializer,
};

use super::{EdgeIndex, NodeIndex};
use crate::{Edge, Graph, Node};

#[derive(Debug)]
enum Error {
    InvalidNode {
        index: usize,
        length: usize,
    },
    InvalidHole {
        index: usize,
    },
    InvalidLength {
        type_: &'static str,
        length: usize,
        max: usize,
    },
    InvalidDirection {
        expected: EdgeProperty,
        received: EdgeProperty,
    },
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::InvalidNode { index, length } => f.write_fmt(format_args!(
                "invalid value: node index `{index}` does not exist in graph with node bound \
                 {length}",
            )),
            Self::InvalidHole { index } => f.write_fmt(format_args!(
                "invalid value: node hole `{index}` is not allowed.",
            )),
            Self::InvalidLength { type_, length, max } => f.write_fmt(format_args!(
                "invalid value: {type_} length `{length}` exceeds maximum of `{max}`",
            )),
            Self::InvalidDirection { expected, received } => f.write_fmt(format_args!(
                "invalid value: expected {expected} graph, but received {received} graph",
            )),
        }
    }
}

impl<T> From<Error> for T
where
    T: serde::de::Error,
{
    fn from(value: Error) -> Self {
        T::custom(value.to_string())
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EdgeProperty {
    Undirected,
    Directed,
}

impl Display for EdgeProperty {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            EdgeProperty::Undirected => f.write_str("undirected"),
            EdgeProperty::Directed => f.write_str("directed"),
        }
    }
}

impl EdgeProperty {
    fn from_type<T>() -> Self
    where
        T: EdgeType,
    {
        if T::is_directed() {
            Self::Directed
        } else {
            Self::Undirected
        }
    }
}

impl<Ix> Serialize for NodeIndex<Ix>
where
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<Ix> Serialize for EdgeIndex<Ix>
where
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<E, Ix> Serialize for Edge<E, Ix>
where
    E: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut tuple = serializer.serialize_tuple(3)?;

        tuple.serialize_element(&self.source())?;
        tuple.serialize_element(&self.target())?;
        tuple.serialize_element(&self.weight)?;

        tuple.end()
    }
}

impl<N, Ix> Serialize for Node<N, Ix>
where
    N: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.weight.serialize(serializer)
    }
}

impl<N, E, Ty, Ix> Serialize for Graph<N, E, Ty, Ix>
where
    N: Serialize,
    E: Serialize,
    Ty: EdgeType,
    Ix: IndexType + Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut struct_ = serializer.serialize_struct("Graph", 4)?;

        struct_.serialize_field("nodes", &self.nodes)?;
        struct_.serialize_field("node_holes", &[])?;
        struct_.serialize_field("edge_property", &EdgeProperty::from_type::<Ty>())?;
        struct_.serialize_field("edges", &self.edges)?;

        struct_.end()
    }
}

#[derive(Deserialize)]
#[serde(rename = "Edge")]
struct EdgeRemote<E, Ix: IndexType>(NodeIndex<Ix>, NodeIndex<Ix>, E);

impl<'de, E, Ix> Deserialize<'de> for Edge<E, Ix>
where
    E: Deserialize<'de>,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let EdgeRemote(source, target, weight) = EdgeRemote::deserialize(deserializer)?;

        Ok(Edge {
            weight,
            node: [source, target],
            next: [EdgeIndex::end(); 2],
        })
    }
}

#[derive(Deserialize)]
#[serde(rename = "Node")]
struct NodeRemote<N>(N);

impl<'de, N, Ix> Deserialize<'de> for Node<N, Ix>
where
    N: Deserialize<'de>,
    Ix: IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let NodeRemote(weight) = NodeRemote::deserialize(deserializer)?;

        Ok(Node {
            weight,
            next: [EdgeIndex::end(); 2],
        })
    }
}

#[derive(Deserialize)]
#[serde(rename = "Graph")]
struct GraphRemote<N, E, Ix: IndexType> {
    nodes: Vec<Node<N, Ix>>,
    // always empty per serialization format
    #[serde(default)]
    node_holes: [(); 0],
    edge_property: EdgeProperty,
    edges: Vec<Edge<E, Ix>>,
}

impl<'de, N, E, Ty, Ix> Deserialize<'de> for Graph<N, E, Ty, Ix>
where
    N: Deserialize<'de>,
    E: Deserialize<'de>,
    Ty: EdgeType,
    Ix: Deserialize<'de> + IndexType,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let GraphRemote {
            nodes,
            node_holes,
            edge_property,
            edges,
        } = GraphRemote::deserialize(deserializer)?;

        if nodes.len() >= <Ix as Integral>::MAX.to_usize() {
            return Err(Error::InvalidLength {
                type_: "node",
                length: nodes.len(),
                max: Ix::MAX.to_usize(),
            }
            .into());
        }

        if edges.len() >= <Ix as Integral>::MAX.to_usize() {
            return Err(Error::InvalidLength {
                type_: "edge",
                length: edges.len(),
                max: Ix::MAX.to_usize(),
            }
            .into());
        }

        let expected = EdgeProperty::from_type::<Ty>();
        if edge_property != expected {
            return Err(Error::InvalidDirection {
                expected,
                received: edge_property,
            });
        }

        let mut this = Graph {
            nodes,
            edges,
            ty: PhantomData,
        };

        let node_count = this.node_count();
        this.link_edges().map_err(|i| Error::InvalidNode {
            index: i.index(),
            length: node_count,
        })?;

        Ok(this)
    }
}
