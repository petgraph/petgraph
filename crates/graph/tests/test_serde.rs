#![cfg(feature = "serde")]

use std::{collections::BTreeSet, fmt::Debug};

use insta::assert_json_snapshot;
use petgraph_core::{
    edge::{Directed, Direction, EdgeType, Undirected},
    index::{DefaultIx, IndexType},
    visit::EdgeRef,
};
use petgraph_graph::{DiGraph, Graph};
use serde::{de::DeserializeOwned, Deserialize};
use serde_json::json;

fn assert_graph_eq<N, N2, E, Ty, Ix>(graph1: &Graph<N, E, Ty, Ix>, graph2: &Graph<N2, E, Ty, Ix>)
where
    N: PartialEq<N2> + Debug,
    N2: PartialEq<N2> + Debug,
    E: PartialEq + Debug,
    Ty: EdgeType,
    Ix: IndexType,
{
    assert_eq!(graph1.node_count(), graph2.node_count());
    assert_eq!(graph1.edge_count(), graph2.edge_count());

    // same node weights
    graph1
        .raw_nodes()
        .iter()
        .zip(graph2.raw_nodes().iter())
        .for_each(|(n1, n2)| assert_eq!(n1.weight, n2.weight));

    // same edge weights
    graph1
        .raw_edges()
        .iter()
        .zip(graph2.raw_edges().iter())
        .for_each(|(n1, n2)| assert_eq!(n1.weight, n2.weight));

    for edge in graph1.edge_references() {
        let (source, target) = graph2.edge_endpoints(edge.id()).expect("edge not found");

        assert_eq!(edge.source(), source);
        assert_eq!(edge.target(), target);
    }

    for index in graph1.node_indices() {
        let outgoing1: BTreeSet<_> = graph1
            .neighbors_directed(index, Direction::Outgoing)
            .collect();
        let outgoing2: BTreeSet<_> = graph2
            .neighbors_directed(index, Direction::Outgoing)
            .collect();

        assert_eq!(outgoing1, outgoing2);

        let incoming1: BTreeSet<_> = graph1
            .neighbors_directed(index, Direction::Incoming)
            .collect();
        let incoming2: BTreeSet<_> = graph2
            .neighbors_directed(index, Direction::Incoming)
            .collect();

        assert_eq!(incoming1, incoming2);
    }
}

fn example<Ty, Ix>() -> Graph<&'static str, i32, Ty, Ix>
where
    Ty: EdgeType,
    Ix: IndexType,
{
    let mut graph = Graph::default();

    let a = graph.add_node("A");
    let b = graph.add_node("B");
    let c = graph.add_node("C");
    let d = graph.add_node("D");
    let e = graph.add_node("E");
    let f = graph.add_node("F");

    graph.extend_with_edges([
        (a, b, 7),
        (c, a, 9),
        (a, d, 14),
        (b, c, 10),
        (d, c, 2),
        (d, e, 9),
        (b, f, 15),
        (c, f, 11),
        (e, f, 6),
    ]);

    // we remove `d` to ensure that holes are handled correctly
    graph.remove_node(d);

    graph
}

#[test]
fn node_str_edges_i32_serialize() {
    let graph: DiGraph<_, _> = example();

    assert_json_snapshot!(&graph);
}

#[test]
fn node_str_edges_i32_deserialize() {
    let graph: DiGraph<_, _> = example();

    let value = serde_value::to_value(graph.clone()).expect("failed to serialize");

    let graph2: DiGraph<String, i32> = DiGraph::deserialize(value).expect("failed to deserialize");

    assert_graph_eq(&graph, &graph2);
}

#[test]
fn node_null_edges_null_serialize() {
    let graph: DiGraph<(), ()> = example().map(|_, _| (), |_, _| ());

    assert_json_snapshot!(&graph);
}

#[test]
fn node_null_edges_null_deserialize() {
    let graph: DiGraph<(), ()> = example().map(|_, _| (), |_, _| ());
    let value = serde_value::to_value(graph.clone()).expect("failed to serialize");

    let graph2: DiGraph<(), ()> = DiGraph::deserialize(value).expect("failed to deserialize");

    assert_graph_eq(&graph, &graph2);
}

#[derive(Debug, Copy, Clone)]
enum Expected {
    Ok,
    Error { error: &'static str },
}

fn assert_inputs<N, E, Ty, Ix>(expected: &[Expected])
where
    N: DeserializeOwned + Debug,
    E: DeserializeOwned + Debug,
    Ty: EdgeType,
    Ix: IndexType + DeserializeOwned,
{
    insta::glob!("inputs/serde/*.json", |path| {
        let content = std::fs::read_to_string(path).expect("failed to read file");
        let index = path
            .file_stem()
            .expect("failed to get file stem")
            .to_string_lossy()
            .split('-')
            .next()
            .expect("failed to get file stem")
            .parse::<usize>()
            .expect("failed to parse file stem");

        let expected = expected[index];

        let graph = serde_json::from_str::<Graph<N, E, Ty, Ix>>(&content);

        match expected {
            Expected::Ok => {
                graph.expect("failed to deserialize");
            }
            Expected::Error {
                error: expected, ..
            } => {
                let error = graph.expect_err("expected error");
                assert_eq!(error.to_string(), *expected);
            }
        }
    });
}

#[test]
fn snapshot_deserialize_default_null() {
    assert_inputs::<(), (), Directed, DefaultIx>(&[
        Expected::Ok,
        Expected::Error {
            error: "invalid value: node index `5` does not exist in graph with length `5`",
        },
        Expected::Error {
            error: "invalid value: node index `300` does not exist in graph with length `5`",
        },
        Expected::Error {
            error: r#"invalid type: string "A", expected unit at line 3 column 7"#,
        },
    ]);
}

#[test]
fn snapshot_deserialize_default_null_undirected() {
    assert_inputs::<(), (), Undirected, DefaultIx>(&[
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
        Expected::Error {
            error: r#"invalid type: string "A", expected unit at line 3 column 7"#,
        },
    ]);
}

#[test]
fn snapshot_deserialize_u8_null() {
    assert_inputs::<(), (), Directed, u8>(&[
        Expected::Ok,
        Expected::Error {
            error: "invalid value: node index `5` does not exist in graph with length `5`",
        },
        Expected::Error {
            error: "invalid value: integer `300`, expected u8 at line 18 column 9",
        },
        Expected::Error {
            error: r#"invalid type: string "A", expected unit at line 3 column 7"#,
        },
    ]);
}

#[test]
fn snapshot_deserialize_default_str_i32() {
    assert_inputs::<String, i32, Directed, DefaultIx>(&[
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Ok,
    ]);
}

#[test]
fn snapshot_deserialize_default_str_i32_undirected() {
    assert_inputs::<String, i32, Undirected, DefaultIx>(&[
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid value: expected undirected graph, but received directed graph",
        },
    ]);
}

#[test]
fn snapshot_deserialize_u8_str_i32() {
    assert_inputs::<String, u8, Directed, DefaultIx>(&[
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Error {
            error: "invalid type: null, expected a string at line 3 column 8",
        },
        Expected::Ok,
    ]);
}

#[test]
fn error_on_too_many_nodes() {
    // The last index (0xFF) is reserved, and cannot be used as a node index.
    let max = u8::MAX;
    let nodes = (0..=max).collect::<Vec<_>>();

    let value = json!({
        "nodes": nodes,
        "edge_property": "directed",
        "edges": []
    });

    let error =
        DiGraph::<u64, (), u8>::deserialize(value).expect_err("expected deserialization to fail");

    assert_eq!(
        error.to_string(),
        "invalid value: node length `256` exceeds maximum of `255`"
    );
}

#[test]
fn error_on_too_many_edges() {
    // The last index (0xFF) is reserved, and cannot be used as an edge index.
    let max = u8::MAX;
    let edges = (0..=max).map(|_| (0, 1, ())).collect::<Vec<_>>();

    let value = json!({
        "nodes": [0, 1],
        "edge_property": "directed",
        "edges": edges
    });

    let error =
        DiGraph::<u64, (), u8>::deserialize(value).expect_err("expected deserialization to fail");

    assert_eq!(
        error.to_string(),
        "invalid value: edge length `256` exceeds maximum of `255`"
    );
}
