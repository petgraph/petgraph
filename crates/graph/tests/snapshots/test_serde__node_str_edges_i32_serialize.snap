---
source: crates/graph/tests/test_serde.rs
expression: "&graph"
---
{
  "nodes": [
    "A",
    "B",
    "C",
    "F",
    "E"
  ],
  "node_holes": [],
  "edge_property": "directed",
  "edges": [
    [
      0,
      1,
      7
    ],
    [
      2,
      0,
      9
    ],
    [
      1,
      3,
      15
    ],
    [
      1,
      2,
      10
    ],
    [
      2,
      3,
      11
    ],
    [
      4,
      3,
      6
    ]
  ]
}
